### ISSUE TITLE
Separate methods for overriding vs calling in bloq counts

### ISSUE THREAD

**Issue by mpharrigan** [quantumlib/Qualtran#373]
It's really useful to have different methods for overriding by bloq authors versus methods designed to be called by a user. For example: A user calls `bloq.decompose_bloq()` which is easy because it has no arguments and a bloq author overrides `build_composite_bloq` which is great because the system can pass in a bloq builder, input soquets, and munge the return value. In fact, we do this for most of our "protocols". Taken from the tutorial notebook:

| What                 | Call this                   | Override this               | Docs                                                    |
|----------------------|-----------------------------|-----------------------------|---------------------------------------------------------|
| Bloq decomposition   | `decompose_bloq()`          | `build_composite_bloq(...)` | [Composite bloq](/_infra/composite_bloq.ipynb)          |
| Numerical simulation | `tensor_contract()`         | `add_my_tensors(...)`       |                                                         |
| Classical simulation | `call_classically(**vals)`  | `on_classical_vals(...)`    | [Classical simulation](/simulation/classical_sim.ipynb) |
| Resource counting    | `bloq_counts()`             | `bloq_counts()`             | [Bloq counts](/resource_counting/bloq_counts.ipynb)     |
| Conversion to Cirq   | `to_cirq_circuit(**quregs)` | `as_cirq_op(...)`           | [Cirq Interop](/cirq_interop/cirq_interop.ipynb)        |

I don't know why I didn't follow this from the start with `bloq_counts` but I want to now.

 - override method is passed a `SympySymbolAllocator` by the library so you can integrate out parameters you don't care about
 - user-callable method lets you pass a generalizer or info on when to stop descending into the call graph.

I need some help with names. It would be nice if you could tell via grammar or something which methods are which flavor, but I don't know exactly what that would look like. So far: the override names are generally more words that contain more prepositions.

Some options

 - `Bloq.get_counts_graph() --> `def my_bloq_counts(self, ssa)`
 - `Bloq.counts_graph()` --> `count_my_bloqs(self, ...)`
 - ??

**Comment by tanujkhattar** [comment]
My recommendation would be to rename to a more general `call_graph` terminology, since that's what everyone was calling it whenever I showed the diagram to new people at IEEE Quantum Week. It's also good to not have the word "bloq" in the name of the protocol so that we can also add the same protocol to Cirq's GateWithRegisters class once the code is moved over the qualtran repo. 

Tl;Dr my suggestion is to have 
- `bloq.call_graph(generalizer)` - User interface that can be called by a user. 
- `def build_call_graph(self, ...)` - Implemented by bloq author.

**Comment by mpharrigan** [comment]
That's reasonable. The only minor issue is that `build_call_graph` won't have anything to do with a graph. It just returns the current node's descendants.

**Comment by tanujkhattar** [comment]
Well, it's part of a protocol that helps build the callgraph and the specific method returns edges corresponding to this node in the callgraph, so I think it should be fine.
